test/distributed/fsdp/test_fsdp_summon_full_params.py

```diff
@@ -0,0 +1,133 @@
+    FSDPTest,
+)
+from torch.testing._internal.common_utils import (
+    TEST_WITH_DEV_DBG_ASAN,
+    run_tests,
+)
+
+
+if not dist.is_available():
+    print("Distributed not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+if TEST_WITH_DEV_DBG_ASAN:
+    print(
+        "Skip dev-asan as torch + multiprocessing spawn have known issues",
+        file=sys.stderr,
+    )
+    sys.exit(0)
+
+
+class TestSummonFullParams(FSDPTest):
+    @property
+    def world_size(self):
+        return 2
+
+    @skip_if_lt_x_gpu(2)
+    def test_summon_full_param_no_writeback(self):
+        model = FSDP(nn.Linear(4, 4).cuda(self.rank))
+        # set the value
+        p = list(model.parameters())[0]
+        with torch.no_grad():
+            # This sets the local shard value
+            p[0] = self.rank + 2
+
+        with model._summon_full_params(volatile=True):
+            with torch.no_grad():
+                p.copy_(torch.zeros_like(p))
+
+        self.assertEqual(p[0], self.rank + 2)
+
+    @skip_if_lt_x_gpu(2)
+    def test_summon_full_param_with_writeback(self):
+        model = FSDP(nn.Linear(3, 3).cuda(self.rank))
+        # set the value
+        p = list(model.parameters())[0]
+        with torch.no_grad():
+            # This sets the local shard value
+            p[0] = self.rank + 2
+
+        with torch.no_grad():
+            with model._summon_full_params(volatile=False):
+                p.copy_(torch.zeros_like(p))
+
+        self.assertEqual(p[0], 0)
+
+    @skip_if_lt_x_gpu(2)
+    def test_summon_full_param_shard_value(self):
+
+        raw_model = nn.Linear(10, 11)
+        # params: 11x10 + 11 -> 121
+        self.assertEqual(121, sum([m.numel() for m in raw_model.parameters()]))
+
+        model = FSDP(raw_model.cuda(self.rank))
+
+        # params should be half as many
+        expected_shard_size = int(math.ceil(121 / self.world_size))
+        self.assertEqual(expected_shard_size, sum([m.numel() for m in model.parameters()]))
+
+        # we're assuming a single flatenned param
+        self.assertEqual(1, len(list(model.parameters())))
+
+        my_shard = torch.clone(next(model.parameters()))
+
+        with model._summon_full_params():
+            # now we see them all
+            self.assertEqual(121, sum([m.numel() for m in model.parameters()]))
+            all_shards = next(model.parameters())
+
+            # this assumes FSDP will chunk the flatenned tensor
+            slice_start = self.rank * expected_shard_size
+            slice_end = (self.rank + 1) * expected_shard_size
+            my_slice = all_shards[slice_start : slice_end]
+            # shards are padded but the full_param tensor is not
+            if self.rank == self.world_size - 1:
+                self.assertTrue(torch.equal(my_shard[0: my_slice.numel()], my_slice))
+            else:
+                self.assertTrue(torch.equal(my_shard, my_slice))
+
+    @skip_if_lt_x_gpu(2)
+    def test_summon_full_param_recursive(self):
```

Comment:
Might be useful to test all these with `CPUOffload` set to true as well to ensure CPU offloading + summon full params work well together. You can use `parametrize` decorator to easily generate those tests.

Reply:
done.